weblate/trans/views/acl.py

```diff
@@ -252,14 +259,60 @@ def manage_access(request, project):
         {
             "object": obj,
             "project": obj,
+            "project_tokens": obj.projecttoken_set.all(),
             "groups": Group.objects.for_project(obj),
             "all_users": User.objects.for_project(obj),
             "blocked_users": obj.userblock_set.select_related("user"),
             "add_user_form": UserManageForm(),
+            "create_project_token_form": ProjectTokenCreateForm(),
             "block_user_form": UserBlockForm(
                 initial={"user": request.GET.get("block_user")}
             ),
             "invite_user_form": InviteUserForm(),
             "ssh_key": get_key_data(),
         },
     )
+
+
+@require_POST
+@login_required
+def delete_token(request, project):
+    """Delete project token."""
+    obj = get_project(request, project)
+
+    if not request.user.has_perm("project.edit", obj):
+        raise PermissionDenied()
+
+    form = ProjectTokenDeleteForm(request.POST)
+
+    if form.is_valid():
+        project_token = get_object_or_404(ProjectToken, pk=form.cleaned_data["token"])
+        if project_token.project.id == obj.id:
+            project_token.delete()
+        else:
+            raise PermissionDenied()
+    else:
+        show_form_errors(request, form)
+
+    return redirect_param("manage-access", "#api", project=obj.slug)
+
+
+@require_POST
+@login_required
+def create_token(request, project):
+    """Create project token."""
+    obj = get_project(request, project)
+
+    if not request.user.has_perm("project.edit", obj):
+        raise PermissionDenied()
+
+    form = ProjectTokenCreateForm(request.POST)
+
+    if form.is_valid():
+        token_value = token_hex()
+        ProjectToken.objects.create(project=obj, token=token_value, **form.cleaned_data)
+        messages.info(request, _("Token has been created: %s") % token_value)
```

Comment:
I'd move the token generating to the model via the default value, see above.

```suggestion
        token = obj.projecttoken_set.create(**form.cleaned_data)
        messages.info(request, _("Token has been created: %s") % token.token)
```

Reply:
Done this as well.